mods/digistuff/rsa.lua

```diff
@@ -0,0 +1,754 @@
+			elseif finalRandom:sub(-1, -1) == "0" then
+				return bigint(finalRandom:sub(1, -2) .. "1")
+			else
+				return bigint(finalRandom)
+			end
+		end
+	end
+end
+
+local function bigRandom(minNum, maxNum)
+	if maxNum < bigint(1000000000) then
+		return bigint(math.random(biginttonumber(minNum),
+			biginttonumber(maxNum)))
+	end
+
+	local maxString = tostring(maxNum)
+	local cap = tonumber(tostring(maxNum):sub(1, 9))
+	local range = #maxString - #tostring(minNum)
+
+	if range == 0 then
+		return bigRandomWithLength(#maxString, cap)
+	end
+
+	if #maxString > 30 then
+		return bigRandomWithLength(#maxString - 1)
+	end
+
+	local randomLength = math.random(1, 2^(#maxString - 1))
+	for i = 1, #maxString - 1 do
+		if randomLength <= (2^i) then
+			return bigRandomWithLength(i)
+		end
+	end
+end
+
+local function isPrime(n)
+	if type(n) == "number" then
+		n = bigint(n)
+	end
+
+	if n % 2 == bigZero then
+		return false
+	end
+
+	local s, d = 0, n - bigOne
+	while d % 2 == bigZero do
+		s, d = s + 1, d / 2
+	end
+
+	for i = 1, 3 do
+		local a = bigRandom(bigint(2), n - 2)
+		local x = modexp(a, d, n)
+		if x ~= bigOne and x + 1 ~= n then
+			for j = 1, s do
+				x = modexp(x, bigint(2), n)
+				if x == bigOne then
+					return false
+				elseif x == n - 1 then
+					a = bigZero
+					break
+				end
+			end
+			if a ~= bigZero then
+				return false
+			end
+		end
+	end
+
+	return true
+end
+
+local function generateLargePrime()
+	local i = 0
+	while true do
+		local randomNumber = bigRandomWithLength(39)
+
+		if isPrime(randomNumber) then
+			return randomNumber
+		end
+	end
+end
+
+local function generatePQ(e)
+	local randomPrime
+	while true do
+		randomPrime = generateLargePrime()
+		if gcd(e, randomPrime - 1) == bigOne then
+			return randomPrime
+		end
+	end
+end
+
+local function euclidean(a, b)
+	local x, y, u, v = bigZero, bigOne, bigOne, bigZero
+	while a ~= bigZero do
+		local q, r = b / a, b % a
+		local m, n = x - u * q, y - v * q
+		b, a, x, y, u, v = a, r, u, v, m, n
+	end
+	return b, x, y
+end
+
+local function modinv(a, m)
+	local gcdnum, x, y = euclidean(a, m)
+	if gcdnum ~= bigOne then
+		return nil
+	else
+		return x % m
+	end
+end
+
+local function generateKeyPair()
+	while true do
+		local e = generateLargePrime()
+		write("-")
+		sleep(0.1)
+		local p = generatePQ(e)
+		write("-")
+		sleep(0.1)
+		local q = generatePQ(e)
+		write("-")
+		sleep(0.1)
+
+		local n = p * q
+		local phi = (p - 1) * (q - 1)
+		local d = modinv(e, phi)
+
+		-- 104328 is just a magic number (can be any semi-unique number)
+		local encrypted = modexp(bigint(104328), e, n)
+		local decrypted = modexp(encrypted, d, n)
+
+		write("+")
+		sleep(0.1)
+		counter = 0
+
+		if decrypted == bigint(104328) then
+			counter = 0
+			return {
+				shared = tostring(n),
+				public = tostring(e),
+			}, {
+				shared = tostring(n),
+				private = tostring(d),
+			}
+		end
+	end
+end
+
+minetest.register_node("digistuff:nic", {
```

Comment:
Why a copy of NIC here?

Reply:
lol oops, I’m unfamiliar with digilines, so I just copied that and changed the code a tad, fixing